File: ios/Sources/FlowbizOnsite/Reachability.swift
import Foundation
import Network

/// Network-restoration seam (SPEC §9 retry trigger). Implementations invoke
/// the callback when connectivity (re)appears; the facade wires it to
/// `FlushController.requestFlush(.networkRestored)` in Slice 4. Injectable
/// so tests use a fake.
protocol ReachabilityMonitor {
    /// Starts monitoring; the callback may fire on the monitor queue. Idempotent.
    func start(onNetworkAvailable: @escaping @Sendable () -> Void)

    /// Stops monitoring. Safe to call when not started.
    func stop()
}

/// Production monitor over `NWPathMonitor` (SPEC §1), running on the SDK's
/// serial queue. Fires only on transitions **to** satisfied — including the
/// initial path update when the network is already up; that one extra flush
/// request is harmless (an empty-queue drain is a no-op).
///
/// Never throws (SPEC §3); a failed monitor degrades to "no reachability
/// trigger" — the other retry triggers still drain the queue.
final class PathMonitorReachability: ReachabilityMonitor, @unchecked Sendable {

    private let queue: DispatchQueue
    private let lock = NSLock()
    private var monitor: NWPathMonitor?
    private var wasSatisfied = false

    /// - Parameter queue: the SDK's serial queue.
    init(queue: DispatchQueue) {
        self.queue = queue
    }

    func start(onNetworkAvailable: @escaping @Sendable () -> Void) {
        lock.lock()
        defer { lock.unlock() }
        guard monitor == nil else { return }
        let pathMonitor = NWPathMonitor()
        pathMonitor.pathUpdateHandler = { [weak self] path in
            guard let self else { return }
            let satisfied = path.status == .satisfied
            self.lock.lock()
            let fire = satisfied && !self.wasSatisfied
            self.wasSatisfied = satisfied
            self.lock.unlock()
            if fire { onNetworkAvailable() }
        }
        pathMonitor.start(queue: queue)
        monitor = pathMonitor
    }

    func stop() {
        lock.lock()
        defer { lock.unlock() }
        monitor?.cancel()
        monitor = nil
        wasSatisfied = false
    }
}

File: android/sdk/src/main/kotlin/com/flowbiz/onsite/Reachability.kt
package com.flowbiz.onsite

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network

/**
 * Network-restoration seam (SPEC §9 retry trigger). Implementations invoke
 * the callback when connectivity (re)appears; the facade wires it to
 * `FlushController.requestFlush(NETWORK_RESTORED)` in Slice 4. Injectable
 * so JVM tests use a fake.
 */
internal interface ReachabilityMonitor {
    /** Starts monitoring; [onNetworkAvailable] may fire on any thread. Idempotent. */
    fun start(onNetworkAvailable: Runnable)

    /** Stops monitoring. Safe to call when not started. */
    fun stop()
}

/**
 * Production monitor over
 * [ConnectivityManager.registerDefaultNetworkCallback] (SPEC §1, available
 * since API 24 — under our API 26 floor). Requires the normal-level
 * `ACCESS_NETWORK_STATE` permission, declared in the SDK manifest and
 * merged into consumers.
 *
 * `onAvailable` also fires once at registration when a network is already
 * up; the resulting extra flush request is harmless (an empty queue drain
 * is a no-op). Never throws: a missing permission or an exhausted callback
 * budget logs and degrades to "no reachability trigger" (SPEC §3) — the
 * other retry triggers still drain the queue.
 */
internal class AndroidReachability(context: Context) : ReachabilityMonitor {

    private val appContext = context.applicationContext
    private val lock = Any()
    private var callback: ConnectivityManager.NetworkCallback? = null

    override fun start(onNetworkAvailable: Runnable) {
        synchronized(lock) {
            if (callback != null) return
            try {
                val manager = appContext.getSystemService(Context.CONNECTIVITY_SERVICE)
                    as? ConnectivityManager ?: return
                val networkCallback = object : ConnectivityManager.NetworkCallback() {
                    override fun onAvailable(network: Network) {
                        try {
                            onNetworkAvailable.run()
                        } catch (_: Throwable) {
                        }
                    }
                }
                manager.registerDefaultNetworkCallback(networkCallback)
                callback = networkCallback
            } catch (t: Throwable) {
                SdkLog.debug("reachability unavailable: ${t.javaClass.simpleName}")
            }
        }
    }

    override fun stop() {
        synchronized(lock) {
            val registered = callback ?: return
            callback = null
            try {
                val manager = appContext.getSystemService(Context.CONNECTIVITY_SERVICE)
                    as? ConnectivityManager ?: return
                manager.unregisterNetworkCallback(registered)
            } catch (t: Throwable) {
                SdkLog.debug("reachability stop failed: ${t.javaClass.simpleName}")
            }
        }
    }
}
